Sphinx nitpicky

asmeurer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

I also want to add something to the documentation guide on how to do cross references, since it's nontrivial to get right especially if you aren't familiar with the RST syntax. With this PR any bad reference will error, so it's important to document how to do them right. 

CC @sympy/gsod 

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

~This is related to~ This fixes https://github.com/sympy/sympy/issues/17619

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
  * Fix various references throughout the docs so that they link properly.
  * Added several missing docstrings to the Sphinx documentation.
* physics.secondquant
  * Added `FockStateFermionBra` and `FockStateFermionKet` to `sympy.physics.secondquant.__all__`.
* integrals
  * Added some documentation on the integration internals to the integration documentation in Sphinx.
<!-- END RELEASE NOTES -->

asmeurer committed:
Fix some RST in the abc docstring

asmeurer committed:
Add Sphinx documentation for the abc module

asmeurer committed:
Add a note about Q to the AssumptionsKeys docstring

asmeurer committed:
Enable the nitpicky flag in Sphinx

This causes the build to warn (and hence error, because we build with -W),
whenever there is a cross reference that doesn't actually link anywhere.

asmeurer commented:
@lglattly here is a draft of some text on how to do cross references in the docs. It applies to both rst and docstrings. This is a bit of a brain dump, so you will probably want to reorganize it a bit. I may comment about more things if they come up as I am working on this. 

## Cross referencing

Any text that references another function should be formatted so that they automatically create a cross reference link to that function's documentation. This is done using the RST cross reference syntax. There are two different kinds of objects that have conventions here:

1. **Objects that are included in `from sympy import *`, for example, `sympy.acos`**. For these: use <code>:obj:\`~.acos()\`</code>. The `~` makes it so that the text in the rendered HTML only shows `acos`. Without it, it would use the fully qualified name `sympy.functions.elementary.trigonometric.acos`. However, for names that are part of the global `sympy` namespace, we don't want to encourage accessing them from their specific submodule, as this is an implementation detail that could change. The `.` makes it so that it finds the function name automatically. Sometimes, Sphinx will give a warning that there are multiple names found. If that happens, replace the `.` with the full name. For example, <code>:obj:\`~sympy.core.basic.Basic.subs()\`</code>. For functions, methods, and classes, it is a convention to add `()` after the name to indicate such. 

You may also use a more specific type indicator instead `obj` (see https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects). However, `obj` will always work, and sometimes SymPy names are not the type you might expect them to be (for example, mathematical function objects such as `sin` are not actually a Python function, rather they are a Python class, so <code>:func:\`~.sin\`</code> will not work).

2. **Objects that are not included in `from sympy import *`, for example, `sympy.physics.vector.dynamicsymbols`.** This can be public API objects from submodules that are not included in the main `sympy/__init__.py`, such as the physics submodule, or private API objects that are not necessarily intended to be used by end-users (but should still be documented). In this case, you must show the fully qualified name, so do not use the `~.` syntax. For example, <code>:obj:\`sympy.physics.vector.dynamicsymbols()\`</code>. 

You may also write custom text that links to the documentation for something using the following syntax <code>:obj:\`custom text\<object>\`</code>. For example, <code>:obj:\`the sine function \<.sin>\`</code>. Note that the `~` character should not be used here. 

Note that references in the "See also" section do not require the `:obj:` syntax. *link to the see also section of the guide here*

If the resulting cross reference is written incorrectly, Sphinx will error when building the docs with an error like

```
WARNING: py:obj reference target not found: expand
```

Here are some troubleshooting tips to fix the errors:

- Make sure you have used the correct syntax, as described above.
- Make sure you spelled the function name correctly.
- Check if the function you are trying to cross reference is actually included in the Sphinx documentation. If it is not, Sphinx will not be able to create a reference for it. In that case, you should add it to the appropriate RST file as described at *link to relevant section of the guide*.
- If the thing you are referring to does not actually have somewhere to link to, do not use the `:obj:` syntax. Rather, simply mark it as code using double backticks. Examples of things that cannot be linked to are Python built in functions like `int` or `NotImplementedError`, functions from other modules outside of SymPy like `matplotlib.plot`, and variable or parameter names that are specific to the text at hand. In general, if the object cannot be accessed as `sympy.something.something.object`, it cannot be cross-referenced and you should not use the `:obj:` syntax. 
- If you are using are using one of the [type specific](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects) identifiers like `:func:`, be sure that the type for it is correct. `:func:` only refers to Python functions. For classes, you need to use `:class:`, and for methods on a class you need to use `:method:`. In general, it is recommended to use `:obj:`, as this will work for any type of object. 
- If you cannot get it to work, go ahead and submit the pull request as is and ask the reviewers for help. 

You may also see errors like

```
WARNING: more than one target found for cross-reference 'subs()': sympy.core.basic.Basic.subs, sympy.matrices.common.MatrixCommon.subs, sympy.physics.vector.vector.Vector.subs, sympy.physics.vector.dyadic.Dyadic.subs
```

for instance, from using <code>:obj:\`~.subs\`</code>. This means that the `.` is not sufficient to find the function, because there are multiple names in SymPy named `subs`. In this case, you need to use the fully qualified name. You can still use `~` to make it shortened in the final text, like <code>:obj:\`~sympy.core.basic.Basic.subs\`</code>.

asmeurer committed:
Fix cross references in gotchas.rst

asmeurer committed:
Add mathjax_config to the docs

See https://github.com/sympy/sphinx-math-dollar/pull/12.

asmeurer commented:
Here's another thing I just noticed that should be mentioned: the line numbers for warnings in Python files are relative to the top of the docstring, not the file itself. Even then the line numbers aren't completely correct. You'll generally have to just search the docstring for the part that the warning is referring to. This is due to a bug in Sphinx.

asmeurer committed:
Fix broken references to sympy.vector in the docs

asmeurer committed:
Fix the name of the SymPyTestResults namedtuple in the test runner

This fixes the Sphinx documentation not being able to find it.

asmeurer committed:
Fix the see also links in the variations() docstring

The functions were removed from compatibility and are not just in the standard
library itertools module. See also doesn't support linking to external docs
directly, so we manually use the Sphinx seealso directive. We could also use
intersphinx, but that would make the build require internet to succeed, and
could potentially break other things that accidentally link to the Python
docs, so I haven't done that for now.

asmeurer committed:
Add ordered() to the Sphinx docs, and fix some RST formatting in it

asmeurer committed:
Replace the documentation for nC, nP, and nT with their docstrings

Also fix some formatting in the docstrings. The formatting should be using the
parameters section, but for now I've just fixed it so it doesn't format
completely wrong.

asmeurer committed:
Make references in See Also sections work without a fully qualified name

As long as they available as sympy.name. In other words, writing 'function' in
a see also section is now the same as if you wrote '~.function' in a normal
Sphinx cross reference. This is only done for references that don't already
have dots.

This way, functions that are in the global SymPy namespace are not referenced
by their fully qualified name, which is preferred as the full module name of a
function that is in the global namespace should not be considered part of the
API (we could move it around without considering it to be an API break).

This fixes #17619.

asmeurer committed:
Fix some more invalid cross references in the docs

This is still far from all of them. I have fixed bad RST formatting for the
handful of functions that I have had to add to the docs for cross references
to work, but I am not fixing any bad formatting outside of that.

asmeurer committed:
Fix some more broken references in the docs

asmeurer committed:
Clean up the formatting in the docstring of rs_series_reversion()

asmeurer committed:
Fix more broken references in the docs

asmeurer committed:
Add documentation for sympy.interactive

init_printing was previously accidentally documented in the physics module
instead of init_vprinting. init_session was not included in Sphinx previously.

asmeurer committed:
Add FockStateFermionBra and FockStateFermionKet to sympy.physics.secondquant.__all__

Otherwise they are not included in the Sphinx docs, but they have the
documentation. Also the aliases FBra and FKet in the Sphinx docs do not cross
reference correctly without this. Since FockBosonBra and FockBosonKet are also
included in __all__ and are similarly aliased (BBra and BKet), it seems
correct that these classes should also be included in __all__.

asmeurer committed:
Fix more bad references in the docs

asmeurer committed:
Add some documentation on the different integration algorithms

I haven't yet fixed all the RST warnings and rendering issues that come from
including these docstrings in Sphinx.

asmeurer committed:
Fix more bad crossreferences in the docs

asmeurer committed:
Fix more bad cross references in the docs

asmeurer committed:
Undo accidental change to the docs Makefile

asmeurer committed:
Fix more bad cross references in the docs

asmeurer commented:
In addition to fixing all the bad cross references, in this PR, I have

- Fixed a few formatting issues in the docs. However, I have kept this to a minimal, because fixing all the formatting issues is a much larger project.
- Added a lot of missing documentation to Sphinx (so that relevant cross references work). I have fixed blatant formatting issues for the docstrings that are new to Sphinx. 
- Made it so that See Alsos use `~.` automatically if they don't contain a `.`, in other words, you don't have to spell out the fully qualified name as long as the name is unique and doesn't have a `.`. See #17619.
- Added some documentation to the integrals.rst about the integration internals, which pulls in the docstrings for all the internal integration functions. It would be good to get a review on this. 
- Added `FockStateFermionBra` and `FockStateFermionKet` to `sympy.physics.secondquant.__all__`. See https://github.com/sympy/sympy/pull/17698/commits/2c1288b4a2525876b903cee4ee2dfd3a48137362

I am almost done fixing all the errors. I will ping here once it is ready to review. I'd like to get this merged ASAP once it is because 1) a PR like this will grow stale very quickly, and 2) I have made some changes to sympy.functions.special docstrings which will affect @lglattly's work in the coming week(s). If anyone wants to review the above additional points now that would be great.

asmeurer committed:
Fix the remaining bad cross references in the docs

asmeurer commented:
So it turns out there weren't that many left. I have fixed all the bad cross references. This is ready for review.

asmeurer commented:
Looks like I need to merge with master to fix some bad references that were added since I started this pull request. That just emphasizes why this should be reviewed and merged quickly.

asmeurer committed:
Merge branch 'master' into sphinx-nitpicky

asmeurer committed:
Fix a broken cross reference in the iproduct docstring

jksuom commented:
manualintegrate is missing from integrals.rst. Its proper place in _eval_integral should probably be before heurisch though it currently comes only afterwards. (Also meijerint_indefinite comes after heurisch in _eval_integral.)

asmeurer commented:
I included it but I forgot to number the paragraph. 

I thought heurisch is called very last. It definitely should be if it isn't.

asmeurer committed:
Fix the paragraph numbering for the integration internals documentation

asmeurer commented:
My changes to have resulted in a "counter too large" error when building the PDF. 

I am also not able to build the pdf myself. I keep getting the error 

```
! Fatal Package fontspec Error: The fontspec package requires either XeTeX or
(fontspec)                      LuaTeX.
(fontspec)
(fontspec)                      You must change your typesetting engine to,
(fontspec)                      e.g., "xelatex" or "lualatex"instead of
(fontspec)                      "latex" or "pdflatex".
```

even though I have installed xelatex (and all the other apt packages that are installed on Travis).

asmeurer committed:
See if moving the meijerint API reference fixes the PDF build

asmeurer committed:
Revert "See if moving the meijerint API reference fixes the PDF build"

This reverts commit 7f2d87d79eff420f2780c77109923c6888bc6c2f.

If it doesn't fix the pdf build, it's not a change we actually want to make.

asmeurer committed:
Fix some invalid LaTeX

asmeurer committed:
Fix the LaTeX docs build

It did not like the g-functions toc being inside of a numbered list, so I
moved it to the main index.rst and just used a basic cross-reference in the
integrals internals documentation.

asmeurer commented:
For some reason I can build the pdf docs with `latexmk -xelatex`. The makefile doesn't seem to want to use -xetex even though it specifically has that in there. 

The issue seems to be with my putting of the g-functions toc inside of a numbered list. I fixed it by moving the g-functions toc into the main index.rst. 

I also noticed that there are even more formatting issues in the PDF docs than there are in the HTML docs, because code blocks do not seem to work. So math that is formatted manually in a code block like in the docstring of `ratint_logpart` does not render properly. It looks like a bug in Sphinx to me, though. 

Anyway, this *should* be ready for a full review now.

asmeurer committed:
Add the --keep-going flag to the docs build

This will make it show all warnings instead of just the first one.

asmeurer commented:
Made one other non-cross reference change: the docs now build with the --keep-going flag so they will show all warnings instead of just the first one.

oscarbenjamin commented on sympy/combinatorics/perm_groups.py:
```diff
@@ -183,7 +183,7 @@ def __getitem__(self, i):
 
     def __contains__(self, i):
-        """Return ``True`` if `i` is contained in PermutationGroup.
+        """Return ``True`` if *i* is contained in PermutationGroup.
```
Why does this use asterisks?

asmeurer replied:
This was a rule that we added in the new style guide based on numpydoc, that parameters are in italics (asterisks). It isn't applied in very many places yet. I think I just edited this one because I was looking at it after adding it to Sphinx.

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -221,17 +221,17 @@ class ComplexRootOf(RootOf):
 
-    >>. r._get_interval()
+    >>> r._get_interval() # doctest: +SKIP
     (-165/169, -206/211)
```
These methods should be internal and not documented I think (not necessarily to be solved in this PR)

asmeurer replied:
It's OK to document internal methods. The documentation can be helpful for developers working on the code. The existence of a docstring or even a doctest shouldn't imply that something is public. 

But we probably should be more particular about what goes in Sphinx. I added a lot of internal things to Sphinx because other things were mentioning them in See Also. I agree we should determine that in a separate issue.

oscarbenjamin replied:
These particular methods are problematic because they can be affected by cache etc. I presume that's why you have `+SKIP` here.

asmeurer replied:
Yes. Actually all I did here was fix the formatting. With `>>.` Sphinx doesn't format it correctly. I made it so the doctest is still not run (I'm assuming the `>>.` was on purpose), but it actually renders correctly.

oscarbenjamin commented:
Looks good to me
